rotkehlchen/chain/evm/decoding/pendle/decoder.py

```diff
@@ -1,122 +1,291 @@
-                    event.event_type = HistoryEventType.WITHDRAWAL
-                    event.event_subtype = HistoryEventSubType.REMOVE_ASSET
-                    event.notes = f'Withdraw {amount} PENDLE from vote escrow'
-                    break
-            else:
-                log.error(f'Could not find pendle unlock transfer for transaction {context.transaction}')  # noqa: E501
+                    event.event_type = HistoryEventType.SPEND
+                    event.event_subtype = HistoryEventSubType.RETURN_WRAPPED
+                    event.notes = f'Return {event.amount} {token_1.symbol} to Pendle'
 
         return DEFAULT_DECODING_OUTPUT
 
+    def _decode_pendle_swap(self, context: DecoderContext) -> DecodingOutput:
+        """This decodes swap events that use the PendleSwap contract.
+
+        Below are aggregators used by PendleSwap with their indexes:
+        Kyberswap -> 1
+        Odos -> 2
+        Okx -> 4
+        1inch -> 5
+
+        Data is gotten from:
+        https://github.com/pendle-finance/pendle-core-v2-public/blob/761033bcc36ecfdb3c523458d7f66aeabb01c3ab/contracts/router/swap-aggregator/IPSwapAggregator.sol#L18
+        """
+        if context.tx_log.topics[0] != SWAP_SINGLE_TOPIC:
+            return DEFAULT_DECODING_OUTPUT
+
+        # Determine swap type and retrieve received token details.
+        # Pendle only emits an event for the token being swapped,
+        # so we must check transaction logs to find the received token
+        amount_out = asset_normalized_value(
+            amount=int.from_bytes(context.tx_log.data[:32]),
+            asset=(token_out := self.base.get_token_or_native(bytes_to_address(context.tx_log.topics[2]))),  # noqa: E501
+        )
+        if (swap_type := int.from_bytes(context.tx_log.topics[1])) == 1:
+            expected_log_topic = KYBERSWAP_SWAPPED_TOPIC
+        elif swap_type == 2:
+            expected_log_topic = ODOS_SWAP_TOPIC
+        elif swap_type == 4:
+            expected_log_topic = OKX_ORDER_RECORD_TOPIC
+        else:  # no transaction was found that touches 1inch so we don't check for it
+            log.debug(f'Found an unsupported swap type {swap_type} in transaction {context.transaction}')  # noqa: E501
+            return DEFAULT_DECODING_OUTPUT
+
+        for tx_log in context.all_logs:
+            if (  # kyberswap
+                    tx_log.topics[0] == expected_log_topic and
+                    bytes_to_address(tx_log.data[96:128]) == PENDLE_ROUTER_ADDRESS
+            ):
+                amount_in = asset_normalized_value(
+                    amount=int.from_bytes(tx_log.data[160:192]),
+                    asset=(token_in := self.base.get_token_or_native(bytes_to_address(tx_log.data[64:96]))),  # noqa: E501
+                )
+                break
+
+            elif (  # odos
+                    tx_log.topics[0] == expected_log_topic and
+                    bytes_to_address(tx_log.data[:32]) == PENDLE_SWAP_ADDRESS
```

Comment:
notice how you are repeating in the 3 branches the topics check.

```suggestion
            if tx_log.topics[0] == expected_log_topic and:
                 continue

            if bytes_to_address(tx_log.data[96:128]) == PENDLE_ROUTER_ADDRESS: # kyberswap
                amount_in = asset_normalized_value(
                    amount=int.from_bytes(tx_log.data[160:192]),
                    asset=(token_in := self.base.get_token_or_native(bytes_to_address(tx_log.data[64:96]))),  # noqa: E501
                )
                break

            elif bytes_to_address(tx_log.data[:32]) == PENDLE_SWAP_ADDRESS:  # odos
```